### ISSUE TITLE
Get different results on different number of GPUs with update-freq set

### ISSUE THREAD

**Issue by Michaelvll** [pytorch/fairseq#1116]
Hi,

I tried to train the transformer base model on WMT'16 EN-DE with the training setting used in lightconv ("PAY LESS ATTENTION WITH LIGHTWEIGHT AND DYNAMIC CONVOLUTIONS "). But I got different results for different number of GPUs.

I trained the model on 8 GPUs, with the settings below:
```yaml
arch: transformer_wmt_en_de
no-progress-bar: true

share-all-embeddings: True
log-interval: 1000
optimizer: adam
adam-betas: (0.9, 0.98)
clip-norm: 0.0
weight-decay: 0.0
criterion: label_smoothed_cross_entropy
label-smoothing: 0.1
min-lr: 1e-09
update-freq: 16

keep-last-epochs: 10
ddp-backend: no_c10d
max-tokens: 4096
lr-scheduler: cosine
warmup-init-lr: 1e-7
warmup-updates: 10000
max-update: 30000
lr-shrink: 1
max-lr: 0.001
lr: 1e-7
min-lr: 1e-9
t-mult: 1
lr-period-updates: 20000

dropout: 0.3
attention-dropout: 0.1

fp16: true
```
I got the BLEU score of 27.49 on the test set.

Then I adapted the setting (update-freq, warmup-updates, max-updates, lr-period-updates), and trained the model on 4 GPUs, with the settings below:
```yaml
arch: transformer_wmt_en_de
no-progress-bar: true

share-all-embeddings: True
log-interval: 100
optimizer: adam
adam-betas: (0.9, 0.98)
clip-norm: 0.0
weight-decay: 0.0
criterion: label_smoothed_cross_entropy
label-smoothing: 0.1
min-lr: 1e-09
update-freq: 32

keep-last-epochs: 10
ddp-backend: no_c10d
max-tokens: 4096
lr-scheduler: cosine
warmup-init-lr: 1e-7
warmup-updates: 20000
max-update: 60000
lr-shrink: 1
max-lr: 0.001
lr: 1e-7
min-lr: 1e-9
t-mult: 1
lr-period-updates: 40000

dropout: 0.3
attention-dropout: 0.1

fp16: true
```
And I got 28.14 with the last checkpoint on the test set which is much higher than the result on 8 GPUs.

I am wondering why the results differ.

Thank you!

**Comment by Michaelvll** [comment]
I got the reason. It seems that I should not change the `max-update`, `warmup-updates` and `lr-period-updates`, since the code call `update-freq` times forward 1 update.

**Comment by netw0rkf10w** [comment]
@Michaelvll Were you able to reproduce the reported 29.xx BLEU scores? I obtained similar result as yours (27.53 vs 27.49), which is far from 29.xx.
